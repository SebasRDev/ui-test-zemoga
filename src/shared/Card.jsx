import { useState } from "react";
import PropTypes from "prop-types";
import "../styles/shared/Card.css";
import { GaugeBar } from "./GaugeBar";
import { Thumb } from "./Thumb";
import { calculateDiferenceDates } from "../helpers/utils";

export const Card = ({ celebrity, handleUpdate }) => {
  const [thumbSelected, setThumbSelected] = useState({
    like: false,
    dislike: false,
  });
  const [voted, setVoted] = useState(false);

  const { name, description, category, picture, lastUpdated, votes } = celebrity;

  const handleSelectVote = (type) => {
    if (type === "up") {
      setThumbSelected({ like: true, dislike: false });
    } else {
      setThumbSelected({ like: false, dislike: true });
    }
  };

  const handleVote = () => {
    if (!voted && (thumbSelected.like || thumbSelected.dislike)){
      setVoted(true);
      const updatedCelebrity = {
        ...celebrity,
        votes: {
          positive: thumbSelected.like ? votes.positive + 1 : votes.positive,
          negative: thumbSelected.dislike ? votes.negative + 1 : votes.negative,
        },
      };
      handleUpdate(updatedCelebrity);
    }
    if (voted) {
      setVoted(false);
      setThumbSelected({ like: false, dislike: false });
    }
  }

  return (
    <div className="voting-card" role="listitem" aria-label={`Voting card ${name}`}>
      <img
        className="voting-card__portrait"
        src={`./assets/img/${picture}`}
        alt={name}
      />
      <div className="voting-card__content">
        <div className="voting-card__data">
          <div className="voting-card__thumb-average" aria-label="Average votes">
            <Thumb type={votes.positive > votes.negative ? "up" : "down"} />
          </div>
          <h4 aria-label="name">{name}</h4>
          <p aria-label="description">{description}</p>
        </div>
        <div className="votin-card__votes">
          {voted ? (
            <p aria-label="Thanks for voting label">Thank you for voting!</p>
          ) : (
            <p aria-label="How many time ago label">
              {calculateDiferenceDates(lastUpdated)} in <span>{category}</span>
            </p>
          )}
          {!voted && (
            <>
              <Thumb
                type="up"
                isSelected={thumbSelected.like}
                handleSelect={handleSelectVote}
                testId={"thumb-up-action"}
              />
              <Thumb
                type="down"
                isSelected={thumbSelected.dislike}
                handleSelect={handleSelectVote}
                testId={"thumb-down-action"}
              />
            </>
          )}
          <button role="button" disabled={!thumbSelected.like && !thumbSelected.dislike} onClick={handleVote} aria-label={`Vote for ${name}`}>
            {voted ? "Vote Again" : "Vote Now"}
          </button>
        </div>
        <GaugeBar votes={votes} />
      </div>
    </div>
  );
};

Card.propTypes = {
  celebrity: PropTypes.object.isRequired,
  handleUpdate: PropTypes.func,
};
