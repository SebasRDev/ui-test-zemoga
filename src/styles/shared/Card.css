.voting-card {
  width: 300px;
  aspect-ratio: 1/1;
  position: relative;
  z-index: 1;
  flex-shrink: 0;
}

.voting-card__thumb-average{
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;

}

.voting-card__portrait {
  position: absolute;
  z-index: -1;
}

.voting-card__content {
  z-index: 1;
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  height: 100%;
  gap: 12px;
}

.voting-card__data {
  color: var(--color-white);
  padding: 0 36px;
  position: relative;

  h4{
    font-size: 30px;
    line-height: 1.2;
    font-weight: normal;
    color: inherit;
    margin: 0;
    margin-bottom: 7px;
  }

  p{
    font-size: 15px;
    line-height: 1.2;
    font-weight: normal;
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    margin: 0;
  }
}

.votin-card__votes{
  display: flex;
  flex-wrap: wrap;
  padding: 0 36px;
  gap: 12px;
  justify-content: flex-end;
  align-items: center;

  p{
    display: block;
    width: 100%;
    font-weight: bold;
    font-size: 12px;
    line-height: 1.2;
    text-align: right;
    color: var(--color-white);
    margin: 0;
  }

  button{
    appearance: none;
    background-color: var(--color-darker-background);
    color: var(--color-white);
    border: 1px solid var(--color-white);
    font-size: 15px;
    line-height: 1.2;
    padding: 10px 20px;
  }
}