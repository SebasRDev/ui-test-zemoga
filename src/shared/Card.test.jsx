import { describe, it, expect, beforeEach, vi } from "vitest";
import {
  render,
  screen,
  act,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import { Card } from "./Card";

const mockData = {
  name: "Kanye West",
  description:
    "Born in Atlanta and raised in Chicago, West was first known as a producer for Roc-A-Fella Records in the early 2000s, producing singles for several mainstream artists.",
  category: "entertainment",
  picture: "kanye.webp",
  lastUpdated: "2024-02-20T23:08:57.892Z",
  votes: {
    positive: 23,
    negative: 36,
  },
};

describe("Card component", () => {
  let handleUpdateMock;
  beforeEach(() => {
    handleUpdateMock = vi.fn();
    render(<Card celebrity={mockData} handleUpdate={handleUpdateMock}/>);
  });
  it("should match the snapshot", () => {
    const { container } = render(<Card celebrity={mockData} />);
    expect(container).toMatchSnapshot();
  });
  it("should render the celebrity name", () => {
    expect(screen.getByText(mockData.name)).toBeInTheDocument();
  });
  it("should render the celebrity description", () => {
    expect(screen.getByText(mockData.description)).toBeInTheDocument();
  });
  it("should render the celebrity category", () => {
    expect(screen.getByText(mockData.category)).toBeInTheDocument();
  });
  it("should render the celebrity picture", () => {
    expect(screen.getByAltText(mockData.name)).toBeInTheDocument();
  });
  it("should have disabled botom when no thumb is selected", () => {
    const button = screen.getByText("Vote Now");
    expect(button).toBeDisabled();
  });
  it("should show active thumb up on click", () => {
    const thumbUp = screen.getByTestId("thumb-up-action");
    const thumbDown = screen.getByTestId("thumb-down-action");
    const voteButton = screen.getByText("Vote Now");
    fireEvent.click(thumbUp);
    expect(thumbUp).toHaveClass("thumb--active");
    expect(thumbDown).not.toHaveClass("thumb--active");
    expect(voteButton).toBeEnabled();
  });
  it("should show active thumb down on click", () => {
    const thumbUp = screen.getByTestId("thumb-up-action");
    const thumbDown = screen.getByTestId("thumb-down-action");
    const voteButton = screen.getByText("Vote Now");
    fireEvent.click(thumbDown);
    expect(thumbDown).toHaveClass("thumb--active");
    expect(thumbUp).not.toHaveClass("thumb--active");
    expect(voteButton).toBeEnabled();
  });
  it("should show Thank you for voting! when vote button is clicked", async () => {
    const thumbUp = screen.getByTestId("thumb-up-action");
    const voteButton = screen.getByText("Vote Now");
    act(() => {
      fireEvent.click(thumbUp);
    });
    fireEvent.click(voteButton);
    await waitFor(() => {
      expect(screen.getByText("Thank you for voting!")).toBeInTheDocument();
      expect(screen.getByText("Vote Again")).toBeInTheDocument();
    });
  });
  it("should reset thumbs and vote button when vote again is clicked", async () => {
    const thumbUp = screen.getByTestId("thumb-up-action");
    const voteButton = screen.getByText("Vote Now");
    act(() => {
      fireEvent.click(thumbUp);
    });
    fireEvent.click(voteButton);
    const voteAgainButton = screen.getByText("Vote Again");
    fireEvent.click(voteAgainButton);
    await waitFor(() => {
      expect(screen.getByText("Vote Now")).toBeInTheDocument();
      expect(screen.getByTestId("thumb-up-action")).not.toHaveClass(
        "thumb--active"
      );
      expect(screen.getByTestId("thumb-down-action")).not.toHaveClass(
        "thumb--active"
      );
    });
  });
  it("should updtate positive votes when vote now is clicked", async () => {
    const thumbUp = screen.getByTestId("thumb-up-action");
    const voteButton = screen.getByText("Vote Now");
    const initialPositiveVotes = mockData.votes.positive;
    act(() => {
      fireEvent.click(thumbUp);
    });
    fireEvent.click(voteButton);
    await waitFor(() => {
      const updatedPositiveVotes = initialPositiveVotes + 1;
      const updatedCelebrity = {
        ...mockData,
        votes: { 
          ...mockData.votes,
          positive: updatedPositiveVotes 
        },
      };
      console.log(handleUpdateMock)
      expect(handleUpdateMock).toHaveBeenCalledWith(updatedCelebrity);
    });
  });
});
