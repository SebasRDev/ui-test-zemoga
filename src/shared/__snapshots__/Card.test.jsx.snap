// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > should match the snapshot 1`] = `
<div>
  <div
    aria-label="Voting card Kanye West"
    class="voting-card"
    role="listitem"
  >
    <img
      alt="Kanye West"
      class="voting-card__portrait"
      src="./assets/img/kanye.webp"
    />
    <div
      class="voting-card__content"
    >
      <div
        class="voting-card__data"
      >
        <div
          aria-label="Average votes"
          class="voting-card__thumb-average"
        >
          <div
            aria-label="thumbs-down"
            class="thumb thumb--down "
            data-testid=""
            role="button"
          >
            <img
              alt="down"
              src="/assets/img/thumbs-down.svg"
            />
          </div>
        </div>
        <h4
          aria-label="name"
        >
          Kanye West
        </h4>
        <p
          aria-label="description"
        >
          Born in Atlanta and raised in Chicago, West was first known as a producer for Roc-A-Fella Records in the early 2000s, producing singles for several mainstream artists.
        </p>
      </div>
      <div
        class="votin-card__votes"
      >
        <p
          aria-label="How many time ago label"
        >
          19 days ago
           in 
          <span>
            entertainment
          </span>
        </p>
        <div
          aria-label="thumbs-up"
          class="thumb thumb--up "
          data-testid="thumb-up-action"
          role="button"
        >
          <img
            alt="up"
            src="/assets/img/thumbs-up.svg"
          />
        </div>
        <div
          aria-label="thumbs-down"
          class="thumb thumb--down "
          data-testid="thumb-down-action"
          role="button"
        >
          <img
            alt="down"
            src="/assets/img/thumbs-down.svg"
          />
        </div>
        <button
          aria-label="Vote for Kanye West"
          disabled=""
          role="button"
        >
          Vote Now
        </button>
      </div>
      <div
        aria-label="Voting gauge"
        class="gauge-bar"
        role="progressbar"
      >
        <div
          class="gauge-bar__data gauge-bar__data--likes"
          style="flex: 1 0 39.0%;"
        >
          <div
            class="gauge-bar__data__content"
          >
            <img
              alt="thumbs-up"
              aria-hidden="true"
              src="./assets/img/thumbs-up.svg"
            />
            <p
              aria-label="Positive percentage"
            >
              39.0%
            </p>
          </div>
        </div>
        <div
          class="gauge-bar__data gauge-bar__data--dislikes"
          style="flex: 1 0 61.0%;"
        >
          <div
            class="gauge-bar__data__content"
          >
            <p
              aria-label="Negative percentage"
            >
              61.0%
            </p>
            <img
              alt="thumbs-down"
              aria-hidden="true"
              src="./assets/img/thumbs-down.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
